mlir/include/mlir/IR/OpImplementation.h

```diff
@@ -1762,6 +1764,15 @@ class OpAsmDialectInterface
                  const SetVector<AsmDialectResourceHandle> &referencedResources,
                  AsmResourceBuilder &builder) const {}
 };
+
+//===--------------------------------------------------------------------===//
+// Custom printers and parsers.
+//===--------------------------------------------------------------------===//
+
+// Handles custom<Shape>(...) in TableGen.
+void printShape(OpAsmPrinter &printer, Operation *op, ArrayRef<int64_t> shape);
+ParseResult parseShape(OpAsmParser &parser, DenseI64ArrayAttr &shape);
+
```

Comment:
Since this is calling `printer.printDimensionList()` I could see it being named `printDimensionList()` as well as a free function here?

Reply:
I renamed it.